### ISSUE TITLE
Intermittent Parameter Error When Switching Display Profiles

### ISSUE THREAD

**Issue by Rtw915** [MartinGC94/DisplayConfig#4]
**Issue Description:**
When running version 1.1.1, I encounter the following error about once per week, which requires me to manually adjust the screens and recreate the profiles.

**Error Details:**
```
PSMessageDetails      : 
Exception             : System.ComponentModel.Win32Exception (0x80004005): The parameter is incorrect
                           at MartinGC94.DisplayConfig.API.DisplayConfig.ApplyConfig(SetDisplayConfigFlags flags)
                           at MartinGC94.DisplayConfig.Commands.UseDisplayConfigCommand.ProcessRecord()
TargetObject          : MartinGC94.DisplayConfig.API.DisplayConfig
CategoryInfo          : InvalidArgument: (MartinGC94.Disp...I.DisplayConfig:DisplayConfig) [Use-DisplayConfig], Win32Exception
FullyQualifiedErrorId : UseDisplayConfigError,MartinGC94.DisplayConfig.Commands.UseDisplayConfigCommand
ErrorDetails          : 
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 5
PipelineIterationInfo : {}

PS C:\WINDOWS\system32> $Error.InvocationInfo


MyCommand             : Use-DisplayConfig
BoundParameters       : {}
UnboundArguments      : {}
ScriptLineNumber      : 5
OffsetInLine          : 46
HistoryId             : 1
ScriptName            : 
Line                  :     Import-Clixml $home\1ScreenProfile.xml | Use-DisplayConfig -AllowChanges
                        
PositionMessage       : At line:5 char:46
                        + ... ort-Clixml $home\1ScreenProfile.xml | Use-DisplayConfig -AllowChanges
                        +                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
PSScriptRoot          : 
PSCommandPath         : 
InvocationName        : Use-DisplayConfig
PipelineLength        : 0
PipelinePosition      : 0
ExpectingInput        : False
CommandOrigin         : Internal
DisplayScriptPosition : 
```

**Script Snippet:**
```powershell
$CurrentDisplayConfig = Get-DisplayConfig

if ((Get-DisplayInfo | Where-Object { $_.Active -eq $true }).Count -ge 3) {
    $CurrentDisplayConfig | Export-Clixml $home\4ScreensProfile.xml
    Import-Clixml $home\1ScreenProfile.xml | Use-DisplayConfig -AllowChanges
} elseif ((Get-DisplayInfo | Where-Object { $_.Active -eq $true }).Count -eq 1) {
    $CurrentDisplayConfig | Export-Clixml $home\1ScreenProfile.xml
    Import-Clixml $home\4ScreensProfile.xml | Use-DisplayConfig -AllowChanges
}
```

**Environment Details:**
- **Operating System:** Windows 11 Home, Version 10.0.22631.3880
- **Display Driver Versions:**
  - NVIDIA GeForce RTX: 31.0.15.5183
  - Intel UHD Graphics: 31.0.101.5187
  - Targus Quad HDMI Video Docking Station: 11.2.3146.0
- **Hardware:** LENOVO 83DF, NVIDIA GeForce RTX 4070
- **Monitors:** 5 monitors connected

**Frequency and Pattern:**
- The error occurs approximately once per week.
- It typically happens when switching between configurations with 5 enabled screens and configurations with 1 screen, and vice versa.

**Steps to Reproduce:**
1. Use the provided PowerShell script to switch display profiles.
2. After a few switches between profiles, the error occurs.

**Temporary Workarounds:**
- Manually adjust the screens and recreate the profiles when the error occurs.

**Question:**
Why is this error occurring?

**Comment by MartinGC94** [comment]
Seems like something changes in your environment that invalidates the old config files. Next time it happens you should manually fix the setup and export a new config file and then compare the old and new config file and see what's different. You can also post the files here and I can take a look at them.

**Comment by Rtw915** [comment]
Good idea! I compared the two files, but I don't quite understand the differences. I had to change the file extensions to upload them to GitHub.

[1ScreenProfile.txt](https://github.com/user-attachments/files/16536471/1ScreenProfile.txt)
[1ScreenProfileOldBusted.txt](https://github.com/user-attachments/files/16536472/1ScreenProfileOldBusted.txt)

**Comment by MartinGC94** [comment]
You are using HDMI. "Forcing" a display is exclusive to analog signals like VGA as far as I can tell, see: https://github.com/MicrosoftDocs/windows-driver-docs/blob/staging/windows-driver-docs-pr/display/forced-versus-connected-targets.md

> All analog target types are considered forceable, and all digital targets are not considered forceable.

I suspect the issue is caused by me not handling your cloned display properly. I remember having to reset some IDs in the `Enable-Display` command code when changing from cloned to extended so it's likely a similar problem here. I'll dig deeper and get back to you in a couple of days.

**Comment by MartinGC94** [comment]
Sorry, I didn't look close enough at your PowerShell code. You need to use the `-UpdateAdapterIds` in `Use-DisplayConfig` to reliably import exported configurations as the adapter LUIDs may change whenever the system is rebooted. `AllowChanges` is also redundant because the config should contain the exact refresh rate values. Try this snippet instead:

```
$CurrentDisplayConfig = Get-DisplayConfig
$DisplayCount = (Get-DisplayInfo | Where-Object -Property Active -EQ $true).Count
if ($DisplayCount -ge 3)
{
    $CurrentDisplayConfig | Export-Clixml $home\4ScreensProfile.xml
    Import-Clixml $home\1ScreenProfile.xml | Use-DisplayConfig -UpdateAdapterIds
} elseif ($DisplayCount -eq 1)
{
    $CurrentDisplayConfig | Export-Clixml $home\1ScreenProfile.xml
    Import-Clixml $home\4ScreensProfile.xml | Use-DisplayConfig -UpdateAdapterIds
}
```

Try it out and let me know how it goes.

**Comment by Rtw915** [comment]
I appreciate your help, but unfortunately, it's now completely broken. I'm encountering the same error in both directions, with only the line number differing.

**Error Details:**

```
PSMessageDetails      : 
Exception             : System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
                           at System.ThrowHelper.ThrowKeyNotFoundException()
                           at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
                           at MartinGC94.DisplayConfig.API.DisplayConfig.UpdateAdapterIds()
                           at MartinGC94.DisplayConfig.Commands.UseDisplayConfigCommand.ProcessRecord()
TargetObject          : MartinGC94.DisplayConfig.API.DisplayConfig
CategoryInfo          : NotSpecified: (MartinGC94.Disp...I.DisplayConfig:DisplayConfig) [Use-DisplayConfig], KeyNotFoundException
FullyQualifiedErrorId : AdapterUpdateError,MartinGC94.DisplayConfig.Commands.UseDisplayConfigCommand
ErrorDetails          : 
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at <ScriptBlock>, C:\Users\Ryan\Documents\Screens.ps1: line 8
PipelineIterationInfo : {}
```

**Comment by MartinGC94** [comment]
The issue you had in your 1ScreenProfile.txt and 1ScreenProfileOldBusted.txt profiles was because the DisplayId order had changed. This was because I didn't take multiple display adapters into account. This has been fixed and you can try it out with the changes I've made in my dev branch: https://github.com/MartinGC94/DisplayConfig/tree/Dev note however that you need to create new profiles from scratch because the old ones will have a different order. Also note that you still need to use the `UpdateAdapterIds` parameter I mentioned before.  
If you don't want to build locally I will also attach the dll to this comment and you can just download and import that one: `import-module C:\PathTo.Dll`.
[DisplayConfig.txt](https://github.com/user-attachments/files/16576057/DisplayConfig.txt)  
rename to .dll and check that the MD5 hash matches `7939B14F1B791E1BE237E2C7AAB985AD`

I cannot reproduce the KeyNotFoundException error you posted most recently but it seems quite strange. The dictionary is built using the available paths so I guess it happens when I'm looping over the ModeArray? https://github.com/MartinGC94/DisplayConfig/blob/main/src/DisplayConfig/API/DisplayConfig.cs#L886 seems weird that the mode array would contain an adapter not in the path array. I could presumably solve this problem by also looping through the ModeArray when creating the old -> new adapter mappings. However, I'd like to get 2 sample profiles where I can see this for myself before I make any changes here. So if you see this error again, make sure to send me a before and after display profile where you've fixed it manually.


So try my most recent changes for a week or two and report back if you experience any issues. If I don't hear anything I will assume it's good and make an official release.

**Comment by MartinGC94** [comment]
My mistake. I forgot that the module had other dependencies outside of the .dll.  
Copy the .dll into your existing displayconfig module folder, replacing the existing .dll. Then import the module like you normally would (either with just the name, or a path to the .psd1).

**Comment by Rtw915** [comment]
Apologies for the delayed response. I encountered some challenges while trying to place the DLL in the appropriate folder due to a UAC permissions issue. After finally managing to copy it over, I'm now facing a new problem when attempting to import the module. Here’s the error I’m receiving:

```
PSMessageDetails      : 
Exception             : System.IO.FileLoadException: Could not load file or assembly 'file:///C:\Program 
                        Files\WindowsPowerShell\Modules\DisplayConfig\1.1.1\DisplayConfig.dll' or one of its dependencies. Operation is not 
                        supported. (Exception from HRESULT: 0x80131515)
                        File name: 'file:///C:\Program Files\WindowsPowerShell\Modules\DisplayConfig\1.1.1\DisplayConfig.dll' ---> 
                        System.NotSupportedException: An attempt was made to load an assembly from a network location which would have 
                        caused the assembly to be sandboxed in previous versions of the .NET Framework. This release of the .NET Framework 
                        does not enable CAS policy by default, so this load may be dangerous. If this load is not intended to sandbox the 
                        assembly, please enable the loadFromRemoteSources switch. See http://go.microsoft.com/fwlink/?LinkId=155569 for more 
                        information.
                           at System.Management.Automation.Runspaces.InitialSessionState.Bind_LoadAssemblies(ExecutionContext context)
                           at System.Management.Automation.Runspaces.InitialSessionState.Bind(ExecutionContext context, Boolean updateOnly, 
                        PSModuleInfo module, Boolean noClobber, Boolean local)
                           at System.Management.Automation.Runspaces.InitialSessionState.Bind(ExecutionContext context, Boolean updateOnly)
                           at Microsoft.PowerShell.Commands.ModuleCmdletBase.LoadModuleManifest(String moduleManifestPath, 
                        ExternalScriptInfo manifestScriptInfo, Hashtable data, Hashtable localizedData, ManifestProcessingFlags 
                        manifestProcessingFlags, Version minimumVersion, Version maximumVersion, Version requiredVersion, Nullable`1 
                        requiredModuleGuid, ImportModuleOptions& options, Boolean& containedErrors)
                        
                        
TargetObject          : 
CategoryInfo          : InvalidOperation: (:) [Import-Module], FileLoadException
FullyQualifiedErrorId : FormatXmlUpdateException,Microsoft.PowerShell.Commands.ImportModuleCommand
ErrorDetails          : 
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}
```

**Comment by MartinGC94** [comment]
A quick google search indicates that the error is caused by the alternative data stream that indicates that the file was downloaded from the internet. Unblock the file with `Unblock-File`, or right click -> Properties -> Unblock.

**Comment by MartinGC94** [comment]
The parameter binding error that mentions PSObject is caused by it not converting the PSObject from `Import-CliXML` into an actual object. There are 2 scenarios where this can happen: 1: You import the .dll directly instead of the .psd1 manifest. 2: You run `Import-CliXML` before importing the module (I'm guessing this is what happened in this scenario).

As for the KeyNotFoundException error, I'd like to compare a working config with a broken one. So when you see it, send me the config file that causes this error as well as a fresh config file created with `Get-DisplayConfig | Export-Clixml`.

**Comment by Rtw915** [comment]
I’m not quite following what’s happening in your scenarios. Here’s what I did:

```powershell
Get-Module
Remove-Module DisplayConfig
Get-Module
# Copied the DLL
Import-Module "C:\Program Files\WindowsPowerShell\Modules\DisplayConfig\1.1.1\DisplayConfig.psd1"
Get-Module
```

However, now the `KeyNotFoundException` error occurs every time, and the script no longer functions as expected.

[1ScreenProfile.txt](https://github.com/user-attachments/files/16643199/1ScreenProfile.txt)

**Comment by MartinGC94** [comment]
I think I have found the cause of the KeyNotFoundException and have added a workaround. As far as I can tell it only happens if you happen to have an adapter that isn't used for any displays. Here's the compiled fix:
[DisplayConfig.txt](https://github.com/user-attachments/files/16646458/DisplayConfig.txt)
Same procedure as before: Download, unblock, check hash and copy into the module folder. MD5 hash: `85B1666F451D651487D04A5FDFF3A575`

**Comment by Rtw915** [comment]
Great news! Everything is working again. For completeness, here's the code I ran that succeeded without needing to recreate the screen profiles from yesterday's attempt:

```powershell
$CurrentDisplayConfig = Get-DisplayConfig
$DisplayCount = (Get-DisplayInfo | Where-Object -Property Active -EQ $true).Count
if ($DisplayCount -ge 3) {
    $CurrentDisplayConfig | Export-Clixml $home\5ScreenProfile.xml
    $DisplayConfig = Import-Clixml $home\1ScreenProfile.xml
    Use-DisplayConfig -DisplayConfig $DisplayConfig -UpdateAdapterIds
} elseif ($DisplayCount -eq 1) {
    $CurrentDisplayConfig | Export-Clixml $home\1ScreenProfile.xml
    $DisplayConfig = Import-Clixml $home\5ScreenProfile.xml
    Use-DisplayConfig -DisplayConfig $DisplayConfig -UpdateAdapterIds
}
```

Thank you so much for resolving this issue and dedicating your time to help. Since the original problem was intermittent, would it be possible to keep this ticket open for 7 days to ensure consistent reliability? Or do you believe the solution has fully addressed the issue?

**Comment by MartinGC94** [comment]
Excellent. I'm going to leave it alone for a week or two and if I don't hear anything from you I'll merge the changes into main and make a new release.

**Comment by MartinGC94** [comment]
This has been fixed in the 2.0 release. I've already published it to the gallery so you can update to the latest version with `Update-Module DisplayConfig`. I'd advice you to also update to this version as it includes some other fixes in addition to the ones you already have.
